File: code/chapter2/pi.py
import multiprocessing as mp
import numpy as np 
from numpy.random import uniform, seed

# first let's implement without parallelization

def count_inside_point(n):
    m = 0
    # we generate n points
    for _ in range(n):
        p_x = uniform(-1, 1)
        p_y = uniform(-1, 1)
        if p_x ** 2 + p_y ** 2 <= 1:
            m += 1
    return m

# now let's try the parallel approach
# each process uses the same seed, which is not desired
def generate_points_parallel(n):
    pool = mp.Pool()
    # we ask each process to generate n//mp.cpu_count() points
    return pool.map(count_inside_point, [n//mp.cpu_count()]*mp.cpu_count())

# set seed for each process
# first, let's define a helper function
def helper(args):
    n, s = args
    seed(s)
    return count_inside_point(n)
    
def generate_points_parallel_set_seed(n):
    pool = mp.Pool() # we can also specify the number of processes by Pool(number)
    # we ask each process to generate n//mp.cpu_count() points
    return pool.map(helper, list(zip([n//mp.cpu_count()]*mp.cpu_count(),range(mp.cpu_count()))))

# another optimization via vectorization
def generate_points_vectorized(n):
    p = uniform(-1, 1, size=(n,2))
    return np.sum(np.linalg.norm(p, axis=1) <= 1)

def pi_naive(n):
    print(f'pi: {count_inside_point(n)/n*4:.6f}')

def pi_parallel(n):
    print(f'pi: {sum(generate_points_parallel_set_seed(n))/n*4:.6f}')

def pi_vectorized(n):
    print(f'pi: {generate_points_vectorized(n)/n*4:.6f}')

File: code/chapter2/pi.R
library(parallel)
count_inside_point = function(n){
  m = 0
  for (i in 1:n){
    p_x = runif(1, -1, 1)
    p_y = runif(1, -1, 1)
    if (p_x^2 + p_y^2 <=1){
      m = m+1
    }
  }
  m
}

# now let's use the mcapply for parallelization
generate_points_parallel = function(n){
  # detectCores() returns the number of cores available
  # we assign the task to each core
  unlist(mclapply(X = rep(n %/% detectCores(), detectCores()),FUN=count_inside_point))
}

# now let's use vectorization
generate_points_vectorized = function(n){
  p = array(runif(n*2,-1,1), c(n,2))
  sum((p[,1]^2+p[,2]^2)<=1)
}

pi_naive = function(n) cat('naive: pi -', 4*count_inside_point(n)/n, '\n')

pi_parallel = function(n) cat('parallel: pi -', 4*sum(generate_points_parallel(n))/n, '\n')

pi_vectorized = function(n) cat('vectorized: pi -', 4*sum(generate_points_vectorized(n))/n, '\n')

